### PR TITLE
update modsec and change wasm repo to v3/dev/wasm-experimental

### PR DESCRIPTION
**Pull request by M4tteoP**
This draft PR is intended to change the modsec wasm repo to the one that has been just pushed under [v3/dev/wasm-experimental](https://github.com/SpiderLabs/ModSecurity/tree/v3/dev/wasm-experimental).
The creation of this new branch led also to update modsecurity from 3.0.4 to 3.0.7.
The library build is working, but afterwards, Envoy is not able to build the module failing with:
```
[2022-06-16 21:22:24.555][1943897][error][wasm] [source/extensions/common/wasm/wasm_vm.cc:38] Failed to load Wasm module due to a missing import: env.getentropy
[2022-06-16 21:22:24.555][1943897][error][wasm] [source/extensions/common/wasm/wasm.cc:109] Wasm VM failed Failed to initialize Wasm code
```
Tested with Envoy version `1.20.1` and `1.21.2`

Any thoughts on that? I'm trying to figure out if it can be fixed tweaking the build process or if tweaks are needed at the ModSecurity codebase.

### CODE REVIEW

**Comment by M4tteoP** [comment]
Refer for example to https://github.com/intel/modsecurity-wasm-filter/pull/17 e2e tests for replicating it

**Comment by jcchavezs** [comment]
Can you try updating the version from 3.0.4 to 3.0.5 and so on to detect where it broke? Maybe easier to open that PR on top of https://github.com/intel/modsecurity-wasm-filter/pull/17

**Comment by M4tteoP** [comment]
Hi! Some updates about it: https://github.com/SpiderLabs/ModSecurity/pull/2763
`getentropy` is internally called by `generate_transaction_unique_id`. I opened the PR reverting the generation of the id back to just the timestamp, but I'm happy to discuss about alternatives avoiding as much as possible not implementing improvements.
Still, it provides an initial fix to have the filter working with the most updated modsecurity

**Comment by bryanwux** [comment]
Update: it works when I use envoy which is built locally. But I got the same error as you (`missing import: env.getentropy`) when I tried to use `func-e` and it said it was using `Envoy version 1.21`. I agree the problem is with `generate_transaction_unique_id` which performs some RNG operations. Could you please try to use envoy instead of `func-e` and see if that works?

**Comment by jcchavezs** [comment]
func-e just pulls envoy generated at
https://github.com/tetratelabs/archive-envoy which is just an envoy build.
What is the envoy version you are compiling? Is it just from source?

søn. 19. jun. 2022, 15:40 skrev Bryan Wu ***@***.***>:

**Comment by bryanwux** [comment]
I am using envoy version 1.22.0-dev. Yes, it is built from source. @jcchavezs

**Comment by M4tteoP** [comment]
An extra information that may help: actually I found out that `getentropy` should have a wasm implementation an is supported by emscripten since version `2.0.5`.
See [release notes ](https://emscripten.org/docs/introducing_emscripten/release_notes.html)looking for it:
```
Add `getentropy` in `sys/random.h`, and use that from libc++'s
  `random_device`. This is more efficient, see #12240.
```
Maybe the limitation is in place from the proxy-sdk stand point?

**Comment by bryanwux** [comment]
It looks like wasm VM does not support /dev/random or /dev/urandom which utilize a hardware entropy source for random numbers. See https://github.com/proxy-wasm/proxy-wasm-cpp-sdk/issues/57. Inside the `generate_transaction_unique_id()`, it calls `random_number()` which defines ` std::random_device rd`, and I think this is the reason why `missing import` appears.

**Comment by bryanwux** [comment]
> It looks like wasm VM does not support /dev/random or /dev/urandom which utilizes a hardware entropy source for random numbers and involves a FS operation. See [proxy-wasm/proxy-wasm-cpp-sdk#57](https://github.com/proxy-wasm/proxy-wasm-cpp-sdk/issues/57). Inside the `generate_transaction_unique_id()`, it calls `random_number()` which defines ` std::random_device rd`, and I think this is the reason why `missing import` appears. Also see: [emscripten-core/emscripten#16896](https://github.com/emscripten-core/emscripten/pull/16896)

**Comment by M4tteoP** [comment]
I see, thanks for the references @bryanwux!
In the end, I think it is like all the other unsupported functions and system calls that rely on the file system: at the moment there is no way to support them and we can just remove/comment it.
I revisited [my commit at Modsecurity PR](https://github.com/SpiderLabs/ModSecurity/pull/2763/commits/19d09a810dabb914463c450a506d3a710b541caf), if you guys are fine with it we can proceed on that way.

**Comment by jcchavezs** [comment]
@M4tteoP could you please update the PR accordingly with this commit https://github.com/SpiderLabs/ModSecurity/pull/2763#issuecomment-1162866585 and see if the e2e gets green?

**Comment by jcchavezs** [comment]
If this gets green we should add a cron for nightly run against latest modsec and detect breaking changes.

**Comment by M4tteoP** [comment]
> @M4tteoP could you please update the PR accordingly with this commit [SpiderLabs/ModSecurity#2763 (comment)](https://github.com/SpiderLabs/ModSecurity/pull/2763#issuecomment-1162866585) and see if the e2e gets green?

I published a branch of this PR: https://github.com/M4tteoP/modsecurity-wasm-filter/tree/remove-getentropy
It just got green :)

**Comment by bryanwux** [comment]
Great! Let's wait for your PR being merged.

**Comment by jcchavezs** [comment]
Sounds good to me.

**Comment by jcchavezs** [comment]
Can we rebase and merge this @M4tteoP ?

File: wasmplugin/Dockerfile [hunk 1]
```diff
@@ -29,10 +29,14 @@ RUN git clone -b v3.9.1 https://github.com/protocolbuffers/protobuf \
     && cd .. \
     && rm -rf protobuf
 
+ARG MODSEC_SHA=a3454df9b09a8de16d41be10cdea4fc46ef08e91
```
Comment (jcchavezs) [inline wasmplugin/Dockerfile:32]: Please do add a comment explaining which repo is this SHA from and maybe also a not on the BRANCH we are aiming?

File: e2e/Makefile [hunk 2]
```diff
@@ -5,7 +5,7 @@ EXTRACT_CONTAINER_NAME=modsecurity-wasm-filter-static-extract
 build-wasm-plugin-static:
 	sed -i 's/envoy-wasm-modsecurity-dynamic/envoy-wasm-modsecurity/' ../wasmplugin/Dockerfile
 	sed -i 's/envoy-wasm-modsecurity-dynamic/envoy-wasm-modsecurity/' ../wasmplugin/Makefile
-	docker build --platform linux/amd64 -t $(IMAGE_NAME):$(IMAGE_VERSION) -f ../wasmplugin/Dockerfile ../wasmplugin
+	docker build --platform linux/amd64 -t $(IMAGE_NAME):$(IMAGE_VERSION) -f ../wasmplugin/Dockerfile ../wasmplugin --build-arg MODSEC_SHA=a3454df9b09a8de16d41be10cdea4fc46ef08e91
```
Comment (jcchavezs) [inline e2e/Makefile:8]: No need to pass a build arg since we have a default value in the docker file.

Comment (M4tteoP) [reply e2e/Makefile]: Okay! I did it to make it ready for further updates, but I removed it :)

File: wasmplugin/Dockerfile (hunk 1 above)
Comment (M4tteoP) [reply wasmplugin/Dockerfile:33]: Yep, done!

**Comment by M4tteoP** [comment]
Done @jcchavezs. Specified the commit based on a build-arg as @bryanwux suggested. I'm going to remove the Draft status

**Comment by bryanwux** [comment]
LGTM. Thank you for working on this.

**Pull request merged**